### DISCUSSION TITLE
Why my rtps emulator didn't work?

### DISCUSSION THREAD

**Discussion by mirusu400** [OpenDDS/OpenDDS#4477] [General]
Hello, for now, I'm testing my emulator works with `DevGuideExamples/DCPS/Messgenger` project. However, even if me continue to emulate publisher behavior, matching and data exchange will not occur normally.

Can anyone check what am I wrong? I fully emulated most submessages, GUIDs, packets. I also minimize my idls for testing.

Here are my fork for test:
https://github.com/mirusu400/OpenDDS/tree/add-logging

Also here are my test python script (need scapy / should manually changing host IP):

[opendds_packet_emulator.zip](https://github.com/OpenDDS/OpenDDS/files/14324721/opendds_packet_emulator.zip)

![image](https://github.com/OpenDDS/OpenDDS/assets/25397908/96217887-e3ee-4de9-a1bd-7a1a07d807f5)

When I turn `-DCPSTransportDebugLevel` to 10, it seems that `DataReaderImpl::transport_assoc_done` didn't triggered (so it cannot get matched), but I didn't find any reasons even I reviewed my packet, scripts, codes.. for almost a week.

Is anyone tell me any wrong points of my script or give me some checking points / tips for debugging 😢

**Comment by mitza-oci** [comment]
Please attach the plaintext log files.

I'm not sure if this is actually a question about OpenDDS - but we can try to answer it anyway.  It might help if you gave some more context about what you're developing.

**Reply by mirusu400** [reply]
[sub_original.txt](https://github.com/OpenDDS/OpenDDS/files/14336032/sub_original.txt)
[sub_emulator2.txt](https://github.com/OpenDDS/OpenDDS/files/14336037/sub_emulator2.txt)

Here are two log files - `sub_original.txt` is connetions subscriber log between both programs - `publisher` and `subscriber`, and `sub_emulator2.txt` is between my `emulator` and `subscriber`

Also here are my running commands:
```
./publisher -DCPSConfigFile rtps.ini -DCPSTransportDebugLevel 10
```

**Comment by jrw972** [comment]
This doesn't seem to be a question about using OpenDDS or how OpenDDS works.
I suggest that you perform a packet capture using an RTPS example, for example, `DevGuideExamples/DCPS/Messenger` with `--rtps`.
You can compare this to the behavior of your emulator.
Then, we can address specific questions about OpenDDS's implementation of RTPS.

**Reply by mirusu400** [reply]
I already capture packets using `DevGuideExamples/DCPS/Messenger`, and made my emulator. I have conducted extensive comparisons, nearly 100 times, between the packets of publisher and my emulator. However, I can't find any difference! I suspect that the challenge in achieving a match may be related to heartbeat and acknack functionalities. Unfortunately, I lack a clear understanding of how these mechanisms operate, hindering my ability to emulate them.

**Reply by jrw972** [reply]
If you do not understand heartbeats and acknacks, then I recommend reading the RTPS Spec https://www.omg.org/spec/DDSI-RTPS/2.5/About-DDSI-RTPS.  Perhaps that will give you the information you need to debug your emulator.
